File: tensorflow/lite/kernels/where.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include <stdint.h>

#include "tensorflow/lite/core/c/c_api_types.h"
#include "tensorflow/lite/core/c/common.h"
#include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
#include "tensorflow/lite/kernels/internal/tensor.h"
#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
#include "tensorflow/lite/kernels/internal/types.h"
#include "tensorflow/lite/kernels/kernel_util.h"

namespace tflite {
namespace ops {
namespace builtin {
namespace where {

constexpr int kInputConditionTensor = 0;
constexpr int kOutputTensor = 0;

template <typename T>
TfLiteStatus ResizeOutputTensor(TfLiteContext* context,
                                const TfLiteTensor* cond_tensor,
                                TfLiteTensor* output_tensor) {
  // Output tensor should have shape:
  // (num_true, cond_rank), where num_true denotes the number of true values
  // in condition.
  const RuntimeShape& cond_shape = GetTensorShape(cond_tensor);
  const int size = cond_shape.FlatSize();
  const int cond_rank = cond_shape.DimensionsCount();
  const T* cond_data = GetTensorData<T>(cond_tensor);

  int true_count = 0;
  for (int i = 0; i < size; ++i) {
    if (cond_data[i] != T(0)) {
      true_count++;
    }
  }
  TfLiteIntArray* output_dims = TfLiteIntArrayCreate(2);
  output_dims->data[0] = true_count;
  output_dims->data[1] = cond_rank;
  return context->ResizeTensor(context, output_tensor, output_dims);
}

template <typename T>
TfLiteStatus PrepareOutput(TfLiteContext* context,
                           const TfLiteTensor* cond_tensor,
                           TfLiteTensor* output) {
  // As output will be a 2D tensor of indices, use int64 to be consistent with
  // tensorflow.
  output->type = kTfLiteInt64;

  // Exit early if cond is a non-const tensor. Set output tensor to dynamic so
  // output size can be determined in Eval.
  if (!IsConstantOrPersistentTensor(cond_tensor)) {
    SetTensorToDynamic(output);
    return kTfLiteOk;
  }
  return ResizeOutputTensor<T>(context, cond_tensor, output);
}

TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
  TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);

  const TfLiteTensor* cond_tensor;
  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, kInputConditionTensor,
                                          &cond_tensor));
  TfLiteTensor* output;
  TF_LITE_ENSURE_OK(context,
                    GetOutputSafe(context, node, kOutputTensor, &output));

  switch (cond_tensor->type) {
    case kTfLiteBool:
      return PrepareOutput<bool>(context, cond_tensor, output);
    case kTfLiteFloat32:
      return PrepareOutput<float>(context, cond_tensor, output);
    case kTfLiteInt64:
      return PrepareOutput<int64_t>(context, cond_tensor, output);
    case kTfLiteInt32:
      return PrepareOutput<int32_t>(context, cond_tensor, output);
    case kTfLiteInt8:
      return PrepareOutput<int8_t>(context, cond_tensor, output);
    case kTfLiteUInt8:
      return PrepareOutput<uint8_t>(context, cond_tensor, output);
    case kTfLiteUInt32:
      return PrepareOutput<uint32_t>(context, cond_tensor, output);
    default:
      TF_LITE_KERNEL_LOG(context,
                         "Condition tensor has unsupported type: '%s'.",
                         TfLiteTypeGetName(cond_tensor->type));
      return kTfLiteError;
  }
  return kTfLiteOk;
}

TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
  const TfLiteTensor* cond_tensor;
  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, kInputConditionTensor,
                                          &cond_tensor));
  TfLiteTensor* output;
  TF_LITE_ENSURE_OK(context,
                    GetOutputSafe(context, node, kOutputTensor, &output));

  if (IsDynamicTensor(output)) {
    switch (cond_tensor->type) {
      case kTfLiteBool:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<bool>(context, cond_tensor, output));
        break;
      case kTfLiteFloat32:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<float>(context, cond_tensor, output));
        break;
      case kTfLiteInt64:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<int64_t>(context, cond_tensor, output));
        break;
      case kTfLiteInt32:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<int32_t>(context, cond_tensor, output));
        break;
      case kTfLiteInt8:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<int8_t>(context, cond_tensor, output));
        break;
      case kTfLiteUInt8:
        TF_LITE_ENSURE_OK(
            context, ResizeOutputTensor<uint8_t>(context, cond_tensor, output));
        break;
      case kTfLiteUInt32:
        TF_LITE_ENSURE_OK(context, ResizeOutputTensor<uint32_t>(
                                       context, cond_tensor, output));
        break;
      default:
        TF_LITE_KERNEL_LOG(context,
                           "Condition tensor has unsupported type: '%s'.",
                           TfLiteTypeGetName(cond_tensor->type));
        return kTfLiteError;
    }
  }

  TfLiteIntArray* dims = cond_tensor->dims;
  if (dims->size == 0) {
    // Scalar tensors are not supported.
    TF_LITE_KERNEL_LOG(context, "Where op requires condition w/ rank > 0");
    return kTfLiteError;
  }

  switch (cond_tensor->type) {
    case kTfLiteBool:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<bool>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteFloat32:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<float>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteInt64:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<int64_t>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteInt32:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<int32_t>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteInt8:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<int8_t>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteUInt8:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<uint8_t>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    case kTfLiteUInt32:
      reference_ops::SelectTrueCoords(GetTensorShape(cond_tensor),
                                      GetTensorData<uint32_t>(cond_tensor),
                                      GetTensorData<int64_t>(output));
      break;
    default:
      TF_LITE_KERNEL_LOG(context,
                         "Condition tensor has unsupported type: '%s'.",
                         TfLiteTypeGetName(cond_tensor->type));
      return kTfLiteError;
  }
  return kTfLiteOk;
}
}  // namespace where

TfLiteRegistration* Register_WHERE() {
  static TfLiteRegistration r = {/*init*/ nullptr, /*free*/ nullptr,
                                 where::Prepare, where::Eval};
  return &r;
}

}  // namespace builtin
}  // namespace ops
}  // namespace tflite

File: tensorflow/lite/testing/op_tests/where.py
# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Test configs for where."""
import tensorflow.compat.v1 as tf
from tensorflow.lite.testing.zip_test_utils import create_tensor_data
from tensorflow.lite.testing.zip_test_utils import make_zip_of_tests
from tensorflow.lite.testing.zip_test_utils import register_make_test_function


@register_make_test_function()
def make_where_tests(options):
  """Make a set of tests to do where."""

  test_parameters = [
      {
          "input_dtype": [tf.float32, tf.int32],
          "input_shape_set": [([1, 2, 3, 4], [1, 2, 3, 4]),],
          "use_where_v2": [False, True],
          "fully_quantize": [False],
      },
      {
          "input_dtype": [tf.float32, tf.int32],
          "input_shape_set": [([], []),],
          "use_where_v2": [],
          "fully_quantize": [False],
      },
      {
          "input_dtype": [tf.float32],
          "input_shape_set": [
              ([1, 2, 3, 4], [1, 2, 3, 4]),
              ([], []),
          ],
          "use_where_v2": [False, True],
          "fully_quantize": [True],
      },
      # High dimension broadcasting support in MLIR converter.
      {
          "input_dtype": [tf.float32, tf.int32],
          "input_shape_set": [([8, 7, 6, 5, 4, 3, 2, 1], [4, 3, 2, 1]),
                              ([8, 7, 6, 5, 4, 3, 2, 1], [None, 3, 2, 1]),
                              ([8, 7, 6, 5, None, 3, 2, 1], [None, 3, 2, 1])],
          "use_where_v2": [True],
          "fully_quantize": [False],
          "dynamic_size_value": [4, 1],
      },
      {
          "input_dtype": [tf.float32],
          "input_shape_set": [([8, 7, 6, 5, 4, 3, 2, 1], [4, 3, 2, 1])],
          "use_where_v2": [True],
          "fully_quantize": [True],
          "dynamic_size_value": [4],
      },
      {
          "input_dtype": [tf.float32, tf.int32],
          "input_shape_set": [([], []), ([1], []), ([], [1])],
          "use_where_v2": [False, True],
          "fully_quantize": [False],
      },
  ]

  def populate_dynamic_shape(parameters, input_shape):
    return [
        parameters["dynamic_size_value"] if x is None else x
        for x in input_shape
    ]

  def build_graph(parameters):
    """Build the where op testing graph."""
    input_value1 = tf.compat.v1.placeholder(
        dtype=parameters["input_dtype"],
        name="input2",
        shape=parameters["input_shape_set"][0])
    input_value2 = tf.compat.v1.placeholder(
        dtype=parameters["input_dtype"],
        name="input3",
        shape=parameters["input_shape_set"][1])
    less = tf.less(input_value1, input_value2)
    where = tf.where_v2 if parameters["use_where_v2"] else tf.where
    out = where(less, input_value1, input_value2)
    return [input_value1, input_value2], [out]

  def build_inputs(parameters, sess, inputs, outputs):
    input_shape_1 = populate_dynamic_shape(parameters,
                                           parameters["input_shape_set"][0])
    input_shape_2 = populate_dynamic_shape(parameters,
                                           parameters["input_shape_set"][1])

    input_value1 = create_tensor_data(
        parameters["input_dtype"], input_shape_1, min_value=-1, max_value=1)
    input_value2 = create_tensor_data(
        parameters["input_dtype"], input_shape_2, min_value=-1, max_value=1)
    return [input_value1, input_value2], sess.run(
        outputs, feed_dict=dict(zip(inputs, [input_value1, input_value2])))

  make_zip_of_tests(
      options,
      test_parameters,
      build_graph,
      build_inputs,
      expected_tf_failures=4)
